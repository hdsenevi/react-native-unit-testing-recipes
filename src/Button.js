import React, { Component } from 'react';
import { Text, TouchableOpacity, Platform, Linking } from 'react-native';

// 1. Changed to a class based component
class Button extends Component {
    constructor(props) {
        super(props);
    }

    // 2. Custom function called onPress TouchableOpacity
    onPressHandler = () => {
        const { onPress, url } = this.props
        if (url) {
            Linking.openURL(url)
        }
        onPress()
    }

    render(){
        const { buttonStyle, textStyle } = styles;
        const { label, primary } = this.props;
        
        // 3. Custom platform specific rendering
        const height = Platform.OS === 'ios' ? 45 : 65;

        // 4. Change color of button depending on 'primary' prop
        const newButtonStyle = primary ? 
            [buttonStyle, { height }] : 
            [buttonStyle, {height, backgroundColor: '#f34541', borderBottomColor: '#a43532'}];

        // 5. Platform specific font styling
        const fontSize = Platform.OS === 'ios' ? 16 : 20;

        return (
            <TouchableOpacity onPress={this.onPressHandler} style={newButtonStyle}>
                <Text style={[textStyle, { fontSize }]}>
                    {label}
                </Text>
            </TouchableOpacity>
        );
    }
};

Button.defaultProps = {
    primary: true,
}

const styles = {
    textStyle: {
        alignSelf: 'center',
        color: '#fff',
        fontWeight: '600',
    },
    buttonStyle: {
        alignSelf: 'stretch',
        justifyContent: 'center',
        backgroundColor: '#38ba7d',
        borderBottomWidth: 6,
        borderBottomColor: '#1e6343',
        borderWidth: 1,
        marginLeft: 15,
        marginRight: 15
    }
};

export default Button;
