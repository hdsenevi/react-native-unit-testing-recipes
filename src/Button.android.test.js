import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

jest.mock('Platform', () => {
    const Platform = require.requireActual('Platform');
    Platform.OS = 'android';
    return Platform;
})

describe('Button', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const component = shallow(<Button label="test label"/>)
            expect(component).toMatchSnapshot()
        });
    });
});
