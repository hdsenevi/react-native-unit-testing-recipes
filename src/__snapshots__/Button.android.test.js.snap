// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering should match to snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    label="test label"
    primary={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <Text
        style={
          Array [
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontWeight": "600",
            },
            Object {
              "fontSize": 20,
            },
          ]
        }
      >
        test label
      </Text>,
      "onPress": [Function],
      "style": Array [
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#38ba7d",
          "borderBottomColor": "#1e6343",
          "borderBottomWidth": 6,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginLeft": 15,
          "marginRight": 15,
        },
        Object {
          "height": 65,
        },
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "test label",
        "style": Array [
          Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontWeight": "600",
          },
          Object {
            "fontSize": 20,
          },
        ],
      },
      "ref": null,
      "rendered": "test label",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <Text
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#fff",
                "fontWeight": "600",
              },
              Object {
                "fontSize": 20,
              },
            ]
          }
        >
          test label
        </Text>,
        "onPress": [Function],
        "style": Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#38ba7d",
            "borderBottomColor": "#1e6343",
            "borderBottomWidth": 6,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginLeft": 15,
            "marginRight": 15,
          },
          Object {
            "height": 65,
          },
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "test label",
          "style": Array [
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontWeight": "600",
            },
            Object {
              "fontSize": 20,
            },
          ],
        },
        "ref": null,
        "rendered": "test label",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
